src/DropdownBase.js

```diff
@@ -0,0 +1,307 @@
+          pullRight={this.props.pullRight}
+          onRequestClose={createChainedFunction(
+            this.props.onRequestClose,
+            this.handleRequestClose
+          )}
+          onSelect={createChainedFunction(
+            this.props.onSelect,
+            this.handleSelect
+          )}>
+          {children}
+        </DropdownMenu>
+      );
+    } else {
+      menuProps.onRequestClose = createChainedFunction(
+        menu.props.onRequestClose,
+        this.props.onRequestClose,
+        this.handleRequestClose
+      );
+      menuProps.onSelect = createChainedFunction(
+        menu.props.onSelect,
+        this.props.onSelect,
+        this.handleSelect
+      );
+      menu = cloneElement(menu, menuProps, menu.props.children);
+    }
+
+    return menu;
+  }
+
+  refineToggle(toggle, children, open) {
+    let toggleProps = {
+      id: this.props.id,
+      ref: TOGGLE_REF,
+      open,
+      bsStyle: this.props.bsStyle,
+      useAnchor: this.useAnchor
+    };
+
+    if (toggle === undefined) {
+      toggle = (
+        <this.Toggle
+          {...toggleProps}
+          onClick={createChainedFunction(
+            this.props[this.Toggle.onClickProp],
+            this.handleClick
+          )}
+          onKeyDown={createChainedFunction(
+            this.props.handleKeyDown,
+            this.handleKeyDown
+          )}
+          noCaret={this.props.noCaret}
+          title={this.props[this.Toggle.titleProp]} />
+      );
+    } else {
+      toggleProps.onClick = createChainedFunction(
+        toggle.props.onClick,
+        this.props[this.Toggle.onClickProp],
+        this.handleClick
+      );
+      toggleProps.onKeyDown = createChainedFunction(
+        toggle.props.handleKeyDown,
+        this.props.handleKeyDown,
+        this.handleKeyDown
+      );
+      toggle = cloneElement(toggle, toggleProps, toggle.props.children);
+    }
+
+    return toggle;
+  }
+}
+
+DropdownBase.Toggle = DropdownToggle;
+DropdownBase.TOGGLE_REF = TOGGLE_REF;
+
+function titleRequired(props, propName, component) {
+  let titles = [];
+
+  if (props.children) {
+    if (props.children instanceof Array) {
+      titles = props.children.filter(child => child.type.isToggle);
+    } else if(props.children.type.isToggle) {
+      titles.push(props.children);
+    }
+  }
+
+  if (titles.length > 1) {
+    return new Error(`(title|children) ${component} - Should only use one ${component}.Toggle child component, only the first ${component}.Toggle will be used`);
+  }
+
+  let title = titles[0];
+
+  if (props.title !== undefined && title !== undefined) {
+    return new Error(`(title|children) ${component} - Must provide either a 'title' prop or a '${component}.Toggle' child, not both.`);
+  }
+
+  if (props.title === undefined && title === undefined) {
+    return new Error(`(title|children) ${component} - Must provide either a 'title' prop or a '${component}.Toggle' child`);
+  }
+}
+
+function childrenAsArray(children) {
```

Comment:
The usage in `menuWithMenuItemSiblingsValidator` looks like it can just be `React.Children.forEach` (or maybe our valid children only one?).

The usage in `singleMenuValidator` seems like it could benefit from e.g. a `ValidComponentChildren.filter`, though.